providers/microsoft/azure/src/airflow/providers/microsoft/azure/hooks/base_azure.py

```diff
@@ -111,22 +119,74 @@ def get_conn(self) -> Any:
             self.log.info("Getting connection using a JSON config.")
             return get_client_from_json_dict(client_class=self.sdk_client, config_dict=key_json)
 
-        credentials: ServicePrincipalCredentials | AzureIdentityCredentialAdapter
+        credentials = self.get_credential(conn=conn)
+
+        return self.sdk_client(
+            credentials=credentials,
+            subscription_id=subscription_id,
+        )
+
+    def get_credential(self, *, conn: Connection | None = None) -> Any:
```

Comment:
Maybe worth tightening the return type here instead of Any, something like a union of the credential classes (ServicePrincipalCredentials | AzureIdentityCredentialAdapter | ClientSecretCredential | DefaultAzureCredential). Not a blocker, but could make the typing clearer.

```suggestion
    def get_credential(self, *, conn: Connection | None = None) -> ServicePrincipalCredentials | AzureIdentityCredentialAdapter | ClientSecretCredential | DefaultAzureCredential:
```

Reply:
updated